File: iotkit-site/internal/semantics/evaluator.go
package semantics

import (
	"errors"
	"math"
)

const maxSafeInteger = int64(9_007_199_254_740_991)

func Evaluate(spec DefinitionSpec, state State, input float64) (Result, State, error) {
	return EvaluateAt(spec, state, input, 0)
}

func EvaluateAt(
	spec DefinitionSpec,
	state State,
	input float64,
	receivedAt int64,
) (Result, State, error) {
	var noResult Result
	if err := spec.Validate(); err != nil {
		return noResult, state, err
	}
	if receivedAt < 0 {
		return noResult, state, errors.New("semantic received time must be non-negative")
	}
	if !finite(input) {
		return noResult, state, errors.New("semantic input must be finite")
	}
	calibrated := input*spec.Scale + spec.Offset
	if !finite(calibrated) {
		return noResult, state, errors.New("calibrated semantic input must be finite")
	}
	result := Result{Calibrated: calibrated}

	if spec.Kind == KindNumeric {
		value := calibrated
		result.Emitted = true
		result.Number = &value
		state.Initialized = true
		return result, state, nil
	}

	active, err := evaluateDetector(spec.Detector, state, calibrated)
	if err != nil {
		return noResult, state, err
	}
	wasInitialized := state.Initialized
	previousActive := state.Active
	if !wasInitialized {
		state.Initialized = true
		state.Active = active
		state.Pending = false
		return emitInitial(spec, result, state)
	}

	if active == state.Active {
		state.Pending = false
	} else {
		debounce := transitionDebounce(spec.Detector, active)
		switch {
		case debounce == 0:
			state.Active = active
			state.Pending = false
		case !state.Pending || state.PendingActive != active:
			state.Pending = true
			state.PendingActive = active
			state.PendingSince = receivedAt
			return result, state, nil
		case receivedAt < state.PendingSince:
			state.PendingSince = receivedAt
			return result, state, nil
		case receivedAt-state.PendingSince < debounce:
			return result, state, nil
		default:
			state.Active = active
			state.Pending = false
		}
	}
	state.Initialized = true
	active = state.Active

	switch spec.Kind {
	case KindBoolean, KindAlarm:
		if previousActive != active {
			value := active
			result.Emitted = true
			result.Boolean = &value
		}
	case KindCumulativeCounter:
		shouldIncrement := spec.Trigger == TriggerActiveSample && active ||
			spec.Trigger == TriggerTransition && !previousActive && active
		if !shouldIncrement {
			return result, state, nil
		}
		if state.Counter >= maxSafeInteger {
			return noResult, state, errors.New("cumulative counter reached the safe integer limit")
		}
		state.Counter++
		value := state.Counter
		result.Emitted = true
		result.Integer = &value
	default:
		return noResult, state, errors.New("unsupported semantic definition kind")
	}
	return result, state, nil
}

func emitInitial(spec DefinitionSpec, result Result, state State) (Result, State, error) {
	switch spec.Kind {
	case KindBoolean, KindAlarm:
		value := state.Active
		result.Emitted = true
		result.Boolean = &value
	case KindCumulativeCounter:
		// The first received value establishes the baseline and is never counted.
	default:
		return Result{}, state, errors.New("unsupported semantic definition kind")
	}
	return result, state, nil
}

func evaluateDetector(detector Detector, state State, value float64) (bool, error) {
	switch detector.Mode {
	case DetectorBooleanHighActive, DetectorBooleanLowActive:
		if value != 0 && value != 1 {
			return false, errors.New("boolean semantic input must be 0 or 1 after correction")
		}
		if detector.Mode == DetectorBooleanHighActive {
			return value == 1, nil
		}
		return value == 0, nil
	case DetectorHighActive:
		if state.Initialized && state.Active {
			return value > detector.FallThreshold, nil
		}
		return value >= detector.RiseThreshold, nil
	case DetectorLowActive:
		if state.Initialized && state.Active {
			return value < detector.RiseThreshold, nil
		}
		return value <= detector.FallThreshold, nil
	default:
		return false, errors.New("unsupported semantic detector")
	}
}

func transitionDebounce(detector Detector, targetActive bool) int64 {
	risingSignal := targetActive
	if detector.Mode == DetectorLowActive ||
		detector.Mode == DetectorBooleanLowActive {
		risingSignal = !targetActive
	}
	if risingSignal {
		return detector.RiseDebounceMS
	}
	return detector.FallDebounceMS
}

func IsSafeInteger(value float64) bool {
	return finite(value) && math.Trunc(value) == value &&
		value >= -float64(maxSafeInteger) && value <= float64(maxSafeInteger)
}

File: edge/src/semantics/evaluator.rs
use serde::{Deserialize, Serialize};

use super::Calibration;

const MAX_DEBOUNCE_MS: i64 = 300_000;
const MAX_SAFE_INTEGER: i64 = 9_007_199_254_740_991;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SemanticKind {
    Numeric,
    Boolean,
    CumulativeCounter,
    Alarm,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DetectorMode {
    #[serde(rename = "")]
    None,
    BooleanHighActive,
    BooleanLowActive,
    HighActive,
    LowActive,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TriggerMode {
    #[serde(rename = "")]
    None,
    OnTransition,
    OnNotification,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct Detector {
    pub mode: DetectorMode,
    pub rise_threshold: f64,
    pub fall_threshold: f64,
    pub rise_debounce_ms: i64,
    pub fall_debounce_ms: i64,
}

impl Default for Detector {
    fn default() -> Self {
        Self {
            mode: DetectorMode::None,
            rise_threshold: 0.0,
            fall_threshold: 0.0,
            rise_debounce_ms: 0,
            fall_debounce_ms: 0,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DefinitionSpec {
    pub kind: SemanticKind,
    pub scale: f64,
    #[serde(default)]
    pub offset: f64,
    #[serde(default)]
    pub detector: Detector,
    #[serde(default = "no_trigger")]
    pub trigger: TriggerMode,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RuleSpec {
    pub kind: SemanticKind,
    #[serde(default)]
    pub detector: Detector,
    #[serde(default = "no_trigger")]
    pub trigger: TriggerMode,
}

const fn no_trigger() -> TriggerMode {
    TriggerMode::None
}

impl DefinitionSpec {
    pub fn validate(self) -> Result<(), SemanticError> {
        Calibration {
            scale: self.scale,
            offset: self.offset,
        }
        .validate()?;
        match self.kind {
            SemanticKind::Numeric
                if self.detector.mode == DetectorMode::None
                    && self.trigger == TriggerMode::None =>
            {
                Ok(())
            }
            SemanticKind::Boolean | SemanticKind::Alarm
                if self.detector.mode != DetectorMode::None
                    && self.trigger == TriggerMode::None =>
            {
                self.detector.validate()
            }
            SemanticKind::CumulativeCounter
                if self.detector.mode != DetectorMode::None
                    && matches!(
                        self.trigger,
                        TriggerMode::OnTransition | TriggerMode::OnNotification
                    ) =>
            {
                self.detector.validate()
            }
            _ => Err(SemanticError::Invalid(
                "semantic kind, detector, and trigger are inconsistent".into(),
            )),
        }
    }
}

impl RuleSpec {
    pub fn validate(self) -> Result<(), SemanticError> {
        DefinitionSpec {
            kind: self.kind,
            scale: 1.0,
            offset: 0.0,
            detector: self.detector,
            trigger: self.trigger,
        }
        .validate()
    }
}

impl Detector {
    fn validate(self) -> Result<(), SemanticError> {
        if !self.rise_threshold.is_finite()
            || !self.fall_threshold.is_finite()
            || !(0..=MAX_DEBOUNCE_MS).contains(&self.rise_debounce_ms)
            || !(0..=MAX_DEBOUNCE_MS).contains(&self.fall_debounce_ms)
        {
            return Err(SemanticError::Invalid(
                "detector thresholds or debounce are invalid".into(),
            ));
        }
        if matches!(
            self.mode,
            DetectorMode::HighActive | DetectorMode::LowActive
        ) && self.fall_threshold > self.rise_threshold
        {
            return Err(SemanticError::Invalid(
                "fall threshold cannot exceed rise threshold".into(),
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct EvaluationState {
    pub initialized: bool,
    pub active: bool,
    pub counter: i64,
    pub pending: bool,
    pub pending_active: bool,
    pub pending_since: i64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Evaluation {
    pub emitted: bool,
    pub number: Option<f64>,
    pub boolean: Option<bool>,
    pub integer: Option<i64>,
    pub calibrated: f64,
}

impl Evaluation {
    fn silent(calibrated: f64) -> Self {
        Self {
            emitted: false,
            number: None,
            boolean: None,
            integer: None,
            calibrated,
        }
    }
}

pub fn evaluate_at(
    spec: DefinitionSpec,
    mut state: EvaluationState,
    input: f64,
    received_at: i64,
) -> Result<(Evaluation, EvaluationState), SemanticError> {
    spec.validate()?;
    if received_at < 0 {
        return Err(SemanticError::Invalid(
            "received time must be non-negative".into(),
        ));
    }
    let calibrated = Calibration {
        scale: spec.scale,
        offset: spec.offset,
    }
    .apply(input)?;
    let mut result = Evaluation::silent(calibrated);
    if spec.kind == SemanticKind::Numeric {
        result.emitted = true;
        result.number = Some(calibrated);
        state.initialized = true;
        return Ok((result, state));
    }

    let detector = spec.detector;
    let candidate = detector_active(detector, state, calibrated)?;
    if !state.initialized {
        state.initialized = true;
        state.active = candidate;
        if matches!(spec.kind, SemanticKind::Boolean | SemanticKind::Alarm) {
            result.emitted = true;
            result.boolean = Some(candidate);
        }
        return Ok((result, state));
    }
    let previous = state.active;
    if candidate == state.active {
        state.pending = false;
    } else {
        let debounce = transition_debounce(detector, candidate);
        if debounce == 0 {
            state.active = candidate;
            state.pending = false;
        } else if !state.pending || state.pending_active != candidate {
            state.pending = true;
            state.pending_active = candidate;
            state.pending_since = received_at;
            return Ok((result, state));
        } else if received_at < state.pending_since {
            state.pending_since = received_at;
            return Ok((result, state));
        } else if received_at - state.pending_since < debounce {
            return Ok((result, state));
        } else {
            state.active = candidate;
            state.pending = false;
        }
    }

    match spec.kind {
        SemanticKind::Boolean | SemanticKind::Alarm if previous != state.active => {
            result.emitted = true;
            result.boolean = Some(state.active);
        }
        SemanticKind::CumulativeCounter => {
            let increments = match spec.trigger {
                TriggerMode::OnNotification => state.active,
                TriggerMode::OnTransition => !previous && state.active,
                TriggerMode::None => unreachable!("validated trigger"),
            };
            if increments {
                if state.counter >= MAX_SAFE_INTEGER {
                    return Err(SemanticError::Invalid(
                        "cumulative counter reached safe integer limit".into(),
                    ));
                }
                state.counter += 1;
                result.emitted = true;
                result.integer = Some(state.counter);
            }
        }
        _ => {}
    }
    Ok((result, state))
}

pub fn evaluate_rule(
    spec: RuleSpec,
    state: EvaluationState,
    calibrated: f64,
    received_at: i64,
) -> Result<(Evaluation, EvaluationState), SemanticError> {
    evaluate_at(
        DefinitionSpec {
            kind: spec.kind,
            scale: 1.0,
            offset: 0.0,
            detector: spec.detector,
            trigger: spec.trigger,
        },
        state,
        calibrated,
        received_at,
    )
}

fn detector_active(
    detector: Detector,
    state: EvaluationState,
    value: f64,
) -> Result<bool, SemanticError> {
    match detector.mode {
        DetectorMode::None => Err(SemanticError::Invalid("detector mode is required".into())),
        DetectorMode::BooleanHighActive | DetectorMode::BooleanLowActive => {
            if value != 0.0 && value != 1.0 {
                return Err(SemanticError::Invalid(
                    "boolean input must be 0 or 1 after calibration".into(),
                ));
            }
            Ok((value == 1.0) == (detector.mode == DetectorMode::BooleanHighActive))
        }
        DetectorMode::HighActive => Ok(if state.initialized && state.active {
            value > detector.fall_threshold
        } else {
            value >= detector.rise_threshold
        }),
        DetectorMode::LowActive => Ok(if state.initialized && state.active {
            value < detector.rise_threshold
        } else {
            value <= detector.fall_threshold
        }),
    }
}

fn transition_debounce(detector: Detector, target_active: bool) -> i64 {
    let rising_signal = if matches!(
        detector.mode,
        DetectorMode::LowActive | DetectorMode::BooleanLowActive
    ) {
        !target_active
    } else {
        target_active
    };
    if rising_signal {
        detector.rise_debounce_ms
    } else {
        detector.fall_debounce_ms
    }
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum SemanticError {
    #[error("invalid semantic configuration: {0}")]
    Invalid(String),
}
